rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,490 @@
+        free: ic::HP as usize,
+    }
+}
+
+#[cfg(feature = "ic")]
+unsafe fn set_limits(limits: &Limits) {
+    use crate::memory::ic;
+    ic::HP = limits.free as u32;
+    ic::LAST_HP = limits.free as u32;
+}
+
+#[cfg(feature = "ic")]
+unsafe fn update_statistics(old_limits: &Limits, new_limits: &Limits) {
+    use crate::memory::ic;
+    let live_size = Bytes(new_limits.free as u32 - new_limits.base as u32);
+    ic::MAX_LIVE = ::core::cmp::max(ic::MAX_LIVE, live_size);
+    ic::RECLAIMED += Bytes(old_limits.free as u64 - new_limits.free as u64);
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum Strategy {
+    Young,
+    Full,
+}
+
+#[cfg(feature = "ic")]
+static mut OLD_GENERATION_THRESHOLD: usize = 32 * 1024 * 1024;
+
+#[cfg(feature = "ic")]
+static mut PASSED_CRITICAL_LIMIT: bool = false;
+
+#[cfg(feature = "ic")]
+const CRITICAL_MEMORY_LIMIT: usize = (4096 - 512) * 1024 * 1024;
+
+#[cfg(feature = "ic")]
+unsafe fn decide_strategy(limits: &Limits) -> Option<Strategy> {
+    const YOUNG_GENERATION_THRESHOLD: usize = 8 * 1024 * 1024;
+
+    let old_generation_size = limits.last_free - limits.base;
+    let young_generation_size = limits.free - limits.last_free;
+
+    if limits.free >= CRITICAL_MEMORY_LIMIT && !PASSED_CRITICAL_LIMIT {
+        PASSED_CRITICAL_LIMIT = true;
+        Some(Strategy::Full)
+    } else if old_generation_size > OLD_GENERATION_THRESHOLD {
+        Some(Strategy::Full)
+    } else if young_generation_size > YOUNG_GENERATION_THRESHOLD {
+        Some(Strategy::Young)
+    } else {
+        None
+    }
+}
+
+#[cfg(feature = "ic")]
+unsafe fn update_strategy(strategy: Strategy, limits: &Limits) {
+    const GROWTH_RATE: f64 = 1.5;
+    if strategy == Strategy::Full {
+        OLD_GENERATION_THRESHOLD = (limits.free as f64 * GROWTH_RATE) as usize;
+        if limits.free < CRITICAL_MEMORY_LIMIT {
+            PASSED_CRITICAL_LIMIT = false
+        }
+    }
+}
+
+pub struct Heap<'a, M: Memory> {
+    pub mem: &'a mut M,
+    pub limits: Limits,
+    pub roots: Roots,
+}
+
+pub struct Roots {
+    pub static_roots: Value,
+    pub continuation_table_ptr_loc: *mut Value,
+}
+
+pub struct Limits {
+    pub base: usize,
+    pub last_free: usize,
+    pub free: usize,
+}
+
+pub struct GenerationalGC<'a, M: Memory> {
+    pub heap: Heap<'a, M>,
+    marked_space: usize,
+    strategy: Strategy,
+}
+
+impl<'a, M: Memory> GenerationalGC<'a, M> {
+    pub fn new(heap: Heap<M>, strategy: Strategy) -> GenerationalGC<M> {
+        GenerationalGC {
+            heap,
+            marked_space: 0,
+            strategy,
+        }
+    }
+
+    pub unsafe fn run(&mut self) {
+        assert_eq!(self.heap.limits.base % 32, 0);
+        let mem_size = Bytes(self.heap.limits.free as u32 - self.heap.limits.base as u32);
+        alloc_bitmap(
+            self.heap.mem,
+            mem_size,
+            self.heap.limits.base as u32 / WORD_SIZE,
+        );
+        alloc_mark_stack(self.heap.mem);
```

Comment:
This would save free reserve for young collection. It will not improve runtime performance, as there is already the `advance()` function in bitmap that is used by the generational GC. And for full collection, it needs this space anyway. I will try this out.

Reply:
No need to try it out if you don't think it's useful. I was just curious. I'll defer to your judgement on all things GC, tbh.

Reply:
Doesn't the bitmap need to be zeroed before use though? That's 256K= 8MiB /4/8 minor heap vs 128 MiB = 4GB/4/8 for a worst case full collection.

Reply:
Maybe best to try out as a stacked PR.... easier to review in isolation. But up to you.

Reply:
Given the current bitmap implementation with the forbidden zone, it might actually be trivial to try out (IIUC): just consider the major heap as part of the static heap when you set up the bitmap.

Reply:
I implemented this. It was indeed a trivial change. Unfortunately, there is no noticeable performance gain. (I used the mark bitmap for an extra sanity check on threading in the old generation to detect multiple threading attempts, but removed this logic now as with remembered sets there should be no duplicates anyway.)

Reply:
That's a shame, sorry for the extra work. I thought this would have showed up as a reduced instruction count for minor gcs.

Reply:
No worries, the change is quite simple and interesting. Probably, it could be beneficial once we have more test cases with larger heaps in the benchmark (with deterministic time slicing).